[TW-1247] API Design Health (#4964)
postmanlabs/postman-docs: src/pages/docs/designing-and-developing-your-api/api-design-health/checking-api-design-health.md

---
title: "Check your API design health"
updated: 2023-08-15
---

> [**API Design Health is available on Postman Free, Basic, and Professional plans.**](https://www.postman.com/pricing)

> You can use [Postman API Design Health](https://go.postman.co/api-health/) to check the design quality of your OpenAPI documents or your [API definitions in Postman](/docs/designing-and-developing-your-api/developing-an-api/defining-an-api/). API Design Health identifies potential security risks your API might be vulnerable to, or potential design issues that might impact its usability. Postman supports API Design Health for OpenAPI documents and API definitions in OpenAPI 3.1, OpenAPI 3.0, and OpenAPI 2.0 format.

Reviewer: Currently we don't support potential security risks. We should remove it for now.

Author: This is fixed.

<!-- screenshot: hero image -->

## Contents

* [About API Design Health](#about-api-design-health)

> * [Add your API to API Design Health](#add-your-api)

Author: I think that's a good idea.

My concern was that "upload" might sound like you're uploading the file to keep it in Postman. And "add" sounded less permanent. But I'm starting to think "upload" is a better term. I'll try to work it in and let you know if I run into any issues.

> * [Add your API to API Design Health](#add-your-api)

Author: This is fixed.

* [Review and fix style guide violations](#review-and-fix-style-guide-violations)
    * [Compare your API with top APIs](#compare-your-api-with-top-apis)
* [Export your updated file](#export-your-updated-file)

## About API Design Health

> Strong API design health means your API design is secure and follows industry best practices. API Design Health checks your OpenAPI document or API definition using a style guide specific to your OpenAPI version, and returns each occurrence of each style guide violation. You then have the option to fix each found violation. To learn more about the style guide and fixing violations, see [Review and fix style guide violations](#review-and-fix-style-guide-violations).

Author: This is fixed.

> API Design Health provides an API design quality score based on each style guide violation it finds. It assigns an impact to each violation and the number of times each violation occurs. This impact is presented in the form of a percentage. A perfect API design quality score is 100%, but the ideal API design quality score for your API may vary. You can also [compare your API's design quality score with top APIs](#compare-your-api-with-top-apis), such as Postman and GitHub. Postman recommends often checking your API's design health during the development process to ensure your API is secure and well-designed.

Author: This is fixed.

## Add your API to API Design Health

To begin checking your API design, you can add your OpenAPI document or API definition file to API Design Health. Then you can review your API design quality score and fix any style guide violations it finds.

1. Go to [Postman API Design Health](https://go.postman.co/api-health/).

>     > If you want to choose a new API, [export the updated OpenAPI document or API definition file](#export-your-updated-file) to download it to your local system. Select **Discard** in the upper-right corner to discard your changes, then choose a new API.

Author: Yes, that's the scenario I'm referring to here. But I just tested the flow where you upload an API definition, and then navigate away from the API Design Health screen. Looks like the file is discarded in this scenario, so I'll remove this note from the docs.

I'll look at possibly relocating this note to a more appropriate section. It might be helpful in case a user wants to check multiple APIs in the same session (meaning before they navigate away from the screen).

>     > If you want to choose a new API, [export the updated OpenAPI document or API definition file](#export-your-updated-file) to download it to your local system. Select **Discard** in the upper-right corner to discard your changes, then choose a new API.

Author: I relocated this content to the Export section. Since the file discards when you navigate away from the screen, it seems like users only need the discard instructions in the context of export.

>     > If you want to choose a new API, [export the updated OpenAPI document or API definition file](#export-your-updated-file) to download it to your local system. Select **Discard** in the upper-right corner to discard your changes, then choose a new API.

Reviewer: > But I just tested the flow where you upload an API definition, and then navigate away from the API Design Health screen. Looks like the file is discarded in this scenario, so I'll remove this note from the docs.

That's not supposed to happen. It's a bug. I will file it.

1. Choose an OpenAPI document or API definition to check. Postman supports both YAML and JSON formats. There are several ways to choose an API:
    * Select **Paste Raw text or URL...**, then enter your OpenAPI document or API definition as raw text or enter a URL.
    * Drag and drop a file, or select a file from your local system.

>     * Select **Choose API from workspace** to choose an API in Postman. Select an API from the dropdown list, then select **Generate API Quality Score**. If you don't see the API you're looking for, make sure you have [Editor or Viewer access](/docs/collaborating-in-postman/roles-and-permissions/#api-roles) to the API.

Author: I addressed the second bullet point. I'll update details related to the first bullet point once the changes are on beta.

>     * Select **Choose API from workspace** to choose an API in Postman. Select an API from the dropdown list, then select **Generate API Quality Score**. If you don't see the API you're looking for, make sure you have [Editor or Viewer access](/docs/collaborating-in-postman/roles-and-permissions/#api-roles) to the API.

Author: "Generate API Quality Score" was recently changed to "Evaluate API". This is now updated in the documentation.

        > To check an API already in Postman, it must have an [API definition](/docs/designing-and-developing-your-api/developing-an-api/defining-an-api/).

    <!-- screenshot: adding screen -->

1. [Review your API's design quality score, and review and fix style guide violations](#review-and-fix-style-guide-violations).

## Review and fix style guide violations

> After you [add your API to API Design Health](#add-your-api-to-api-design-health), you can review and fix style guide violations in your OpenAPI document or API definition. Each style guide violation shows the impact it has on your design quality score, the number of times each violation occurred, and the severity level of each style guide violation. You can expand each guideline to review each occurrence of each violation, and learn more about how to fix it. You can also [compare your API with top APIs](#comparing-your-api-with-top-apis) to help you identify your ideal design quality score.

Author: Is "OpenAPI document" the preferred term? It looks like this is the term we use in the API Governance Postman docs. Also, it looks like OpenAPI's definition of "OpenAPI document" matches what we're referring to. Please see https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#openapi-document

I might be mistaken. What do you think?

> After you [add your API to API Design Health](#add-your-api-to-api-design-health), you can review and fix style guide violations in your OpenAPI document or API definition. Each style guide violation shows the impact it has on your design quality score, the number of times each violation occurred, and the severity level of each style guide violation. You can expand each guideline to review each occurrence of each violation, and learn more about how to fix it. You can also [compare your API with top APIs](#comparing-your-api-with-top-apis) to help you identify your ideal design quality score.

Reviewer: Ah, thanks for the reference. This is okay then.

> After you [add your API to API Design Health](#add-your-api-to-api-design-health), you can review and fix style guide violations in your OpenAPI document or API definition. Each style guide violation shows the impact it has on your design quality score, the number of times each violation occurred, and the severity level of each style guide violation. You can expand each guideline to review each occurrence of each violation, and learn more about how to fix it. You can also [compare your API with top APIs](#comparing-your-api-with-top-apis) to help you identify your ideal design quality score.

Author: Great!

> To fix each occurrence of a style guide violation, hover over the occurrence and select **Auto fix**. API Design Health will insert a template into your OpenAPI document or API definition file in the editor. Fill out the template to fix the style guide violation. To learn more about the style guide and how to fix each violation, see the style guide relevant to your OpenAPI version: [OpenAPI 3.1](/docs/designing-and-developing-your-api/api-design-health/openapi-31-style-guide/), [OpenAPI 3.0](/docs/designing-and-developing-your-api/api-design-health/openapi-30-style-guide/), or [OpenAPI 2.0](/docs/designing-and-developing-your-api/api-design-health/openapi-20-style-guide/).

Author: Thanks! I'll update these details when the changes are on beta.

> To fix each occurrence of a style guide violation, hover over the occurrence and select **Auto fix**. API Design Health will insert a template into your OpenAPI document or API definition file in the editor. Fill out the template to fix the style guide violation. To learn more about the style guide and how to fix each violation, see the style guide relevant to your OpenAPI version: [OpenAPI 3.1](/docs/designing-and-developing-your-api/api-design-health/openapi-31-style-guide/), [OpenAPI 3.0](/docs/designing-and-developing-your-api/api-design-health/openapi-30-style-guide/), or [OpenAPI 2.0](/docs/designing-and-developing-your-api/api-design-health/openapi-20-style-guide/).

Author: This is now updated.

In the upper right of the editor are options to change the file format, wrap text, copy, search, and undo your last change.

<!-- screenshot: select fix -->

> Postman can return new style guide violations when you manually edit your OpenAPI document or API definition in the editor.

When you finish editing your API definition, [export your updated OpenAPI document or API definition](#export-your-updated-file) to save your changes.

> If you're checking an API already in Postman, edits aren't automatically saved to your API definition file in Postman.

### Comparing your API with top APIs

You can compare your API with top APIs, such as Postman and GitHub. Select **Compare with top APIs** to display a table comparing your API with several top APIs.

> You can compare design quality score, the total number of style guide violations, and the number of times each API violates several curated guidelines. You can use this information to help you identify the ideal design quality score for your API.

Author: Thanks! I'll update these details when the changes are on beta.

> You can compare design quality score, the total number of style guide violations, and the number of times each API violates several curated guidelines. You can use this information to help you identify the ideal design quality score for your API.

Author: This is now updated.

---
title: "Check your API design health"
updated: 2023-08-15
---

> You can use [Postman API Design Health](https://go.postman.co/api-design-health/) to check the design quality of your OpenAPI documents or your [API definitions in Postman](/docs/designing-and-developing-your-api/developing-an-api/defining-an-api/). API Design Health identifies potential design issues that might impact its usability. Postman supports API Design Health for OpenAPI documents and API definitions in OpenAPI 3.1, OpenAPI 3.0, and OpenAPI 2.0 format.

Reviewer 2: ```suggestion
You can use [Postman API Design Health](https://go.postman.co/api-design-health/) to check the design quality of your OpenAPI documents or your [API definitions in Postman](/docs/designing-and-developing-your-api/developing-an-api/defining-an-api/). API Design Health identifies potential design issues that might impact your API's usability. Postman supports API Design Health for OpenAPI documents and API definitions in OpenAPI 3.1, OpenAPI 3.0, and OpenAPI 2.0 format.
```

Author: Agreed. This is much better.

    * Select **Paste Raw text or URL...**, then enter your OpenAPI document or API definition as raw text or enter a URL.
    * Drag and drop a file, or select a file from your local system.
    * Select **Choose API from workspace** to choose an API in Postman. Select an API from the dropdown list, then select **Evaluate API**. If you don't see the API you're looking for, make sure you have [Editor access](/docs/collaborating-in-postman/roles-and-permissions/#api-roles) to the API.

        > To check an API already in Postman, it must have an [API definition](/docs/designing-and-developing-your-api/developing-an-api/defining-an-api/).

    <!-- screenshot: adding screen -->

1. [Review your API's design quality score, and review and fix style guide violations](#review-and-fix-style-guide-violations).

## Review and fix style guide violations

You can review and fix style guide violations in your OpenAPI document or API definition. API Design Health checks your OpenAPI document or API definition using a style guide specific to your OpenAPI version, and returns occurrences of each style guide violation. You then have the option to fix each found violation.

### Review violations

> After you [upload your API to API Design Health](#upload-your-api-to-api-design-health), you can review style guide violations in your OpenAPI document or API definition. Each style guide violation shows the impact it has on your design quality score, the number of times each violation occurred, and the severity level of each style guide violation. The severity level of style guide violations from most severe to least severe is Error, Warning, Info, and Hint.

Reviewer 2: ```suggestion
After you [upload your API to API Design Health](#upload-your-api-to-api-design-health), you can review style guide violations in your OpenAPI document or API definition. Each style guide violation shows the impact it has on your design quality score, the number of times each violation occurred, and the severity level of each style guide violation. The severity levels of style guide violations from most severe to least severe are Error, Warning, Info, and Hint.
```

In the upper right of the editor are options to change the file format, wrap text, copy, search, and undo your last change. <!-- might relocate this information -->

To fix style guide violations, do the following:

1. Expand a style guide violation to review each occurrence of the violation.
1. Hover over the occurrence of the style guide violation, then select **Fix**. Your OpenAPI document or API definition will scroll to the location of the violation, adding or removing fields depending on the violation.

    <!-- screenshot: selecting fix button -->

1. You might need to edit the fix for the style guide violation, depending on the violation. If fields are added, Postman recommends editing placeholder values with details relevant to your API.

    > Postman can return new style guide violations when you manually edit your OpenAPI document or API definition in the editor.

> 1. You can review the impact each fix has on your design quality score and total number of violations. As you edit your OpenAPI definition or API definition, a number displays to the right of your design quality score and total number of violations, showing how your edits increase or decrease these numbers.

Reviewer 2: ```suggestion
1. You can review the impact each fix has on your design quality score and total number of violations. As you edit your OpenAPI document or API definition, a number displays to the right of your design quality score and total number of violations, showing how your edits increase or decrease these numbers.
```

Author: Thanks for catching this!

## Review and fix style guide violations

You can review and fix style guide violations in your OpenAPI document or API definition. API Design Health checks your OpenAPI document or API definition using a style guide specific to your OpenAPI version, and returns occurrences of each style guide violation. You then have the option to fix each found violation.

### Review violations

> After you [upload your API to API Design Health](#upload-your-api-to-api-design-health), you can review style guide violations in your OpenAPI document or API definition. Each style guide violation shows the impact it has on your design quality score, the number of times each violation occurred, and the severity level of each style guide violation. The severity level of style guide violations from most severe to least severe is Error, Warn, and Hint.

Reviewer: There is INFO also if I am not wrong.

Author: I'm not seeing guidelines in the modal with the INFO tag. I'll check with the group.

Author: This is now updated.

You can expand each guideline to review each occurrence of a violation, and learn more about how to fix it. Select the occurrence of the violation to scroll to its location in your OpenAPI document or API definition. You can also [compare your API with top APIs](#compare-your-api-with-top-apis) to help you identify your ideal design quality score.

<!-- screenshot: showing rule violations and how to expand -->

Once you're done reviewing your style guide violations, you can [fix violations](#fix-violations) in your OpenAPI document or API definition.

#### Compare your API with top APIs

> You can compare your API with top APIs, such as Postman and GitHub. Select **Compare** to display a table comparing your API with several top APIs. You can select different top APIs using the dropdown menu to compare APIs relevant to your API.

Reviewer: I would recommend not using our own API as a reference. Lets select another one.

Author: This is now updated: https://github.com/postmanlabs/postman-docs/pull/4964/commits/33fa68b4fd5897910fa2da219fc5c528359e98a7
